src/libaktualizr/uptane/uptanerepository.cc

```diff
@@ -117,113 +100,93 @@ bool Repository::verifyMetaTargets(const Uptane::Targets& director_targets, cons
   return true;
 }
 
-Json::Value Repository::getJSON(const std::string& url) {
-  HttpResponse response = http.get(url);
-  if (!response.isOk()) {
-    throw MissingRepo(url);
-  }
-  return response.getJson();
-}
-
-/**
- * Note this doesn't check the validity of version numbers, only that the signed part is actually signed.
- * @param repo
- * @param role
- * @param version
- * @param root_used
- * @return The contents of the "signed" section
- */
-Json::Value Repository::fetchRole(const std::string& base_url, Uptane::Role role, Version version) {
-  // TODO: chain-loading root.json
-  return getJSON(base_url + "/" + version.RoleFileName(role));
-}
-
-bool Repository::getMeta() {
-  MetaPack new_meta;
+bool Repository::feedCheckMeta() {
   TimeStamp now(TimeStamp::Now());
+  MetaPack new_meta;
+  RawMetaPack meta_raw;
 
-  new_meta.director_root =
-      Uptane::Root(now, "director", fetchRole(config.uptane.director_server, Role::Root()), meta_.director_root);
-  new_meta.image_root = Uptane::Root(now, "repo", fetchRole(config.uptane.repo_server, Role::Root()), meta_.image_root);
-
-  new_meta.image_timestamp = Uptane::TimestampMeta(
-      now, "repo", fetchRole(config.uptane.repo_server, Role::Timestamp(), Version()), meta_.image_root);
-  if (new_meta.image_timestamp.version() > meta_.image_timestamp.version()) {
-    new_meta.image_snapshot = Uptane::Snapshot(
-        now, "repo", fetchRole(config.uptane.repo_server, Role::Snapshot(), Version()), meta_.image_root);
-    new_meta.image_targets = Uptane::Targets(
-        now, "repo", fetchRole(config.uptane.repo_server, Role::Targets(), Version()), meta_.image_root);
-  } else {
-    new_meta.image_snapshot = meta_.image_snapshot;
-    new_meta.image_targets = meta_.image_targets;
+  // should be already put to the storage by fetcher
+  if (!storage->loadUncheckedMetadata(&meta_raw)) {
+    LOG_WARNING << "No metadata to check";
+    return false;
   }
 
-  new_meta.director_targets = Uptane::Targets(
-      now, "director", fetchRole(config.uptane.director_server, Role::Targets(), Version()), meta_.director_root);
-  if (new_meta.director_root.version() > meta_.director_root.version() ||
-      new_meta.image_root.version() > meta_.image_root.version() ||
-      new_meta.director_targets.version() > meta_.director_targets.version() ||
-      new_meta.image_timestamp.version() > meta_.image_timestamp.version()) {
-    if (verifyMetaTargets(new_meta.director_targets, new_meta.image_targets)) {
-      storage->storeMetadata(new_meta);
-      meta_ = new_meta;
-    } else {
-      LOG_WARNING << "Metadata image/directory repo consistency check failed.";
+  try {
```

Comment:
Loading of each piece of metadata should be its own function. Also, this fails to check things like

- the `targets` version in `snapshot` matches the actual version in the `targets` metadata
- the `snapshot` version in `timestamp` matches the actual version in the `snapshot` metadata
- and others